site/docs/tutorial/sql_to_substrait.md

```diff
@@ -0,0 +1,995 @@
+    Substrait will define it's own human-readable text format, but for now this
+    tutorial will make due with what Protobuf provides.
+
+Substrait is designed to communicate plans (mostly logical plans). Those plans
+contain types, schemas, expressions, extensions, and relations. We'll look at 
+them in that order, going from simplest to most complex until we can construct
+full plans.
+
+![](substrait_components.svg)
+
+This tutorial won't cover all the details of each piece, but it will give you an
+idea of how they connect together. For a detailed reference of each individual 
+field, the best place to look is reading the 
+[protobuf definitions](https://github.com/substrait-io/substrait/tree/main/proto/substrait).
+They represent the source-of-truth of the spec and are well-commented to address
+ambiguities.
+
+## Problem Set up
+
+To learn Substrait, we'll build up to a specific query. We'll be using the
+tables:
+
+```sql
+CREATE TABLE orders (
+  product_id: i64 NOT NULL,
+  quantity: i32 NOT NULL,
+  order_date: date NOT NULL,
+  price: decimal(10, 2)
+);
+```
+
+```sql
+CREATE TABLE products (
+  product_id: i64 NOT NULL,
+  categories: list<string NOT NULL> NOT NULL,
+  details: struct<manufacturer: string, year_created: int32>,
+  product_name: string
+);
+```
+
+This `orders` table represents events where products were sold, recording how
+many (`quantity`) and at what price (`price`). The `products` table provides
+details for each product, with `product_id` as the primary key.
+
+And we'll try to create the query:
+
+```sql
+SELECT
+  product_name,
+  product_id,
+  sum(quantity * price) as sales
+FROM
+  orders
+INNER JOIN
+  products
+ON
+  orders.product_id = products.product_id
+WHERE
+  -- categories does not contain "Computers"
+  INDEX_IN("Computers", categories) IS NULL
+GROUP BY
+  product_name,
+  product_id
+```
+
+The query asked the question: For products that aren't in the `"Computer"`
+category, how much has each product generated in sales?
+
+However, Substrait doesn't correspond to SQL as much as it does to logical plans.
+So to be less ambiguous, the plan we are aiming for looks like:
+
+```
+|-+ Aggregate({sales = sum(quantity_price)}, group_by=(product_name, product_id))
+  |-+ InnerJoin(on=orders.product_id = products.product_id)
+    |- ReadTable(orders)
+    |-+ Filter(INDEX_IN("Computers", categories) IS NULL)
+      |- ReadTable(products)
+```
+
+## Types and Schemas
+
+As part of the Substrait plan, we'll need to embed the data types of the input
+tables. In Substrait, each type is a distinct message, which at a minimum 
+contains a field for nullability. For example, a string field looks like:
+
+```json
+{
+  "string": {
+    "nullability": "NULLABILITY_NULLABLE"
+  }
+}
+```
+
+Nullability is an enum not a boolean, since Substrait allows 
+`NULLABILITY_UNSPECIFIED` as an option, in addition to `NULLABILITY_NULLABLE`
+(nullable) and `NULLABILITY_REQUIRED` (not nullable).
```

Comment:
Eh...generally `_UNSPECIFIED` is an invalid value for an enum and will be rejected by all plans and consumers.  Pretty much every enum has an `_UNSPECIFIED` variant, even when it doesn't make sense (e.g. `JOIN_TYPE_UNSPECIFIED`).  This is done whenever an enum doesn't have a natural default value because otherwise there is no way in protobuf to distinguish between "use the default" and "this is missing".  So I'm not entirely sure why nullability is a boolean.  @jacques-n is `NULLABILITY_UNSPECIFIED` meant to be a legitimate thing?  I would think, if someone doesn't know, they would just pick `NULLABILITY_NULLABLE`.